homeassistant/components/teslemetry/services.py

```diff
@@ -0,0 +1,312 @@
+                vol.Optional(ATTR_ENABLE): bool,
+                vol.Optional(ATTR_PRECONDITIONING_ENABLED): bool,
+                vol.Optional(ATTR_PRECONDITIONING_WEEKDAYS): bool,
+                vol.Optional(ATTR_DEPARTURE_TIME): str,
+                vol.Optional(ATTR_OFF_PEAK_CHARGING_ENABLED): bool,
+                vol.Optional(ATTR_OFF_PEAK_CHARGING_WEEKDAYS): bool,
+                vol.Optional(ATTR_END_OFF_PEAK_TIME): str,
+            }
+        ),
+    )
+
+    async def valet_mode(call: ServiceCall) -> None:
+        """Configure fleet telemetry."""
+        device = async_get_device_for_service_call(hass, call)
+        config = async_get_config_for_device(hass, device)
+        vehicle = async_get_vehicle_for_entry(hass, device, config)
+
+        await wake_up_vehicle(vehicle)
+        await handle_vehicle_command(
+            vehicle.api.set_valet_mode(
+                call.data.get("enable"), call.data.get("pin", "")
+            )
+        )
+
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_VALET_MODE,
+        valet_mode,
+        schema=vol.Schema(
+            {
+                vol.Required(CONF_DEVICE_ID): cv.string,
+                vol.Required(ATTR_ENABLE): cv.boolean,
+                vol.Required(ATTR_PIN): All(cv.positive_int, Range(min=1000, max=9999)),
+            }
+        ),
+    )
+
+    async def speed_limit(call: ServiceCall) -> None:
+        """Configure fleet telemetry."""
+        device = async_get_device_for_service_call(hass, call)
+        config = async_get_config_for_device(hass, device)
+        vehicle = async_get_vehicle_for_entry(hass, device, config)
+
+        await wake_up_vehicle(vehicle)
+        enable = call.data.get("enable")
+        if enable is True:
+            await handle_vehicle_command(
+                vehicle.api.speed_limit_activate(call.data.get("pin"))
+            )
+        elif enable is False:
+            await handle_vehicle_command(
+                vehicle.api.speed_limit_deactivate(call.data.get("pin"))
+            )
+
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_SPEED_LIMIT,
+        speed_limit,
+        schema=vol.Schema(
+            {
+                vol.Required(CONF_DEVICE_ID): cv.string,
+                vol.Required(ATTR_ENABLE): cv.boolean,
+                vol.Required(ATTR_PIN): All(cv.positive_int, Range(min=1000, max=9999)),
+            }
+        ),
+    )
+
+    async def time_of_use(call: ServiceCall) -> None:
+        """Configure time of use settings."""
+        device = async_get_device_for_service_call(hass, call)
+        config = async_get_config_for_device(hass, device)
+        site = async_get_energy_site_for_entry(hass, device, config)
+
+        resp = await handle_command(
+            site.api.time_of_use_settings(call.data.get(ATTR_TOU_SETTINGS))
+        )
+        if "error" in resp:
+            raise ServiceValidationError(resp["error"])
```

Comment:
errors should be `HomeAssistantError`. Also please use translations.

Reply:
This error has unknown values, so I cant translate it.

Reply:
You can pass it as a placeholder so the string could look something like `Service returned error: {error}` or something similar to make it a bit more user friendly than just the hard errors

Reply:
Oh yeah good idea.